File: php/src/ErrorDebugger.php
<?php
namespace ErrorDebugger;

use OpenTelemetry\SDK\Trace\TracerProvider;
use OpenTelemetry\SDK\Trace\SpanProcessor\BatchSpanProcessorBuilder;
use OpenTelemetry\SDK\Resource\ResourceInfo;
use OpenTelemetry\SDK\Resource\ResourceInfoFactory;
use OpenTelemetry\Contrib\Otlp\SpanExporter;
use OpenTelemetry\Contrib\Otlp\OtlpHttpTransportFactory;

class ErrorDebugger
{
    private static bool $initialized = false;
    private static TracerProvider $tracerProvider;
    
    public static function init(Config $config): void
    {
        if (self::$initialized) return;
        
        $resource = ResourceInfoFactory::defaultResource()->merge(ResourceInfo::create(
            \OpenTelemetry\SDK\Common\Attribute\Attributes::create([
                'service.name' => $config->serviceName,
                'deployment.environment' => $config->environment,
                'project.token' => $config->token,
                'telemetry.sdk.language' => 'php',
                'telemetry.sdk.name' => 'errordebugger-php'
            ])
        ));
        
        $endpoint = $config->dsn . (str_ends_with($config->dsn, '/') ? 'v1/otlp/traces' : '/v1/otlp/traces');
        
        $transport = (new OtlpHttpTransportFactory())->create(
            $endpoint,
            'application/x-protobuf',
            ['x-project-token' => $config->token]
        );
        
        $exporter = new SpanExporter($transport);
        
        self::$tracerProvider = new TracerProvider(
            (new BatchSpanProcessorBuilder($exporter))->build(),
            null,
            $resource
        );
        
        self::$initialized = true;
    }
    
    public static function captureException(\Throwable $exception): void
    {
        if (!self::$initialized) return;

        $tracer = self::$tracerProvider->getTracer('errordebugger');
        $span = $tracer->spanBuilder('manual_exception')->startSpan();
        $span->recordException($exception);
        $span->setStatus(\OpenTelemetry\API\Trace\StatusCode::STATUS_ERROR);
        $span->end();
    }
    
    public static function shutdown(): void
    {
        if (self::$initialized) {
            self::$tracerProvider->shutdown();
        }
    }
}

File: ruby/lib/errordebugger.rb
require 'opentelemetry/sdk'
require 'opentelemetry/exporter/otlp'
require 'errordebugger/config'

module ErrorDebugger
  @initialized = false
  @tracer_provider = nil

  def self.init(config)
    return if @initialized

    resource = OpenTelemetry::SDK::Resources::Resource.create({
      'service.name' => config.service_name,
      'deployment.environment' => config.environment,
      'project.token' => config.token,
      'telemetry.sdk.language' => 'ruby',
      'telemetry.sdk.name' => 'errordebugger-ruby'
    })

    endpoint = config.dsn.end_with?('/') ? "#{config.dsn}v1/otlp/traces" : "#{config.dsn}/v1/otlp/traces"

    exporter = OpenTelemetry::Exporter::OTLP::Exporter.new(
      endpoint: endpoint,
      headers: { 'x-project-token' => config.token }
    )

    @tracer_provider = OpenTelemetry::SDK::Trace::TracerProvider.new(
      resource: resource
    )
    @tracer_provider.add_span_processor(
      OpenTelemetry::SDK::Trace::Export::BatchSpanProcessor.new(exporter)
    )

    OpenTelemetry.tracer_provider = @tracer_provider
    @initialized = true
  end

  def self.capture_exception(exception)
    # OpenTelemetry returns a NOOP tracer if not initialized, so this is safe
    tracer = OpenTelemetry.tracer_provider.tracer('errordebugger')
    span = tracer.start_span('manual_exception')
    span.record_exception(exception)
    span.status = OpenTelemetry::Trace::Status.error
    span.finish
  end
  
  # For testing
  def self.reset
    @initialized = false
    @tracer_provider = nil
  end
end
